lib/rmt/file_validator.rb

```diff
@@ -0,0 +1,48 @@
+require 'fileutils'
+
+class RMT::FileValidator
+  class << self
+    def validate_local_file(file_reference, deep_verify:)
+      if valid_on_disk?(file_reference, deep_verify)
+        DownloadedFile.track_file(checksum: file_reference.checksum,
+                                  checksum_type: file_reference.checksum_type,
+                                  local_path: file_reference.local_path,
+                                  size: file_reference.size)
+        return true
+      end
+
+      # Remove invalid files/DB entries as soon as they are found
+      FileUtils.remove_file(file_reference.local_path, force: true)
+      DownloadedFile.where(local_path: file_reference.local_path).destroy_all
+      false
+    end
+
+    def find_valid_files_by_checksum(checksum, checksum_type, deep_verify:)
+      valid_files_on_disk = DownloadedFile
+        .where(checksum: checksum, checksum_type: checksum_type)
+        .group_by { |f| valid_on_disk?(f, deep_verify) }
+
+      # Remove invalid files/DB entries as soon as they are found
+      valid_files_on_disk[false]&.each do |file|
+        FileUtils.remove_file(file.local_path, force: true)
+        file.destroy
+      end
+
+      valid_files_on_disk[true] || []
+    end
```

Comment:
```suggestion
    def find_valid_files_by_checksum(checksum, checksum_type, deep_verify:)
      files = DownloadedFile.where(checksum: checksum, checksum_type: checksum_type)

      files.delete_if do |file|
        next false if valid_on_disk?(f, deep_verify)

        FileUtils.remove_file(file.local_path, force: true)
        file.destroy
        true
      end
    end
```

Reply:
Nice shorter version! I just needed to add `.to_a` after the `.where(...)`, I guess `ActiveRecord::Relation` implements the `Enumerable` interface but not the `Array` one; `delete_if` is specific for arrays. I'm going to commit this change together with the remaining others.

Reply:
Done :heavy_check_mark: